Don't migrate technical users

geropl opened the pull request:
## Description

With https://github.com/gitpod-io/gitpod/pull/17672 we started using another trigger (websocket creation) for starting user migration, which also migrated technical users. This path did not take into account the "single org" constraint, that's why login/setup flows started failing on Dedicated installations.

Because a) "technical user" accounts as well as b) the "org-only migration" are temporary constructs until we have better solutions, this PR makes sure that invariant holds again by:
- blocking technical users from being migrated
- delete all organizations that have been created for those technical users

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes WEB-431, WEB-409

## How to test
- create a new preview env based on this PR
- try to follow the Dedicated Setup flow
- should result in
![image](https://github.com/gitpod-io/gitpod/assets/32448529/441ac17d-a97d-4060-937d-d2910ab28223)


## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - gpl-fix-sso-flow</li>
	<li><b>🔗 URL</b> - <a href="https://gpl-fix-sso-flow.preview.gitpod-dev.com/workspaces" target="_blank">gpl-fix-sso-flow.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [x] with-dedicated-emulation
- [x] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>
 
/hold

easyCZ commented on components/gitpod-db/src/typeorm/migration/1685447855319-RevertOrgOnlyMigrationForBuiltinUser.ts:
```diff
@@ -0,0 +1,38 @@
+        // Delete the create organizations
+        // We need to go by name because due to another bug, we might have created n organizations
+        queryRunner.query(
+            "DELETE FROM d_b_team WHERE name = 'agent-smith' OR name = 'builtin-workspace-prober' or name = 'admin-user'",
```
[follow-up] Do we still really need these users?

geropl replied:
We need them until we can replace them with system-delivered service accounts (AFAIK).

easyCZ commented on components/gitpod-db/src/typeorm/migration/1685447855319-RevertOrgOnlyMigrationForBuiltinUser.ts:
```diff
@@ -0,0 +1,38 @@
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        // Unset all user values set by UserToTeamMigrationService
+        queryRunner.query(
+            "UPDATE d_b_user SET usageAttributionId = '', additionalData = JSON_REMOVE(additionalData, '$.isMigratedToTeamOnlyAttribution', '$.shouldSeeMigrationMessage') WHERE id IN ('builtin-user-workspace-probe-0000000', 'builtin-user-agent-smith-0000000', 'f071bb8e-b5d1-46cf-a436-da03ae63bcd2')",
```
Interesting that these are not actually valid UUIDs, but that's secondary to this PR

easyCZ commented on components/gitpod-db/src/typeorm/migration/1685447855319-RevertOrgOnlyMigrationForBuiltinUser.ts:
```diff
@@ -0,0 +1,38 @@
+            "DELETE FROM d_b_team WHERE name = 'agent-smith' OR name = 'builtin-workspace-prober' or name = 'admin-user'",
+        );
+        queryRunner.query(
+            "DELETE FROM d_b_team_membership WHERE teamId IN (SELECT id FROM d_b_team WHERE name = 'agent-smith' OR name = 'builtin-workspace-prober' or name = 'admin-user')",
```
If this were to execute at the time when the `admin-user` is being used to setup an Org, it will break the setup flow

geropl replied:
I don't think so: This is limited to the point orgs named "admin-user", which only happens during this migration.

easyCZ replied:
Ah, gotcha. Sorry I missed that this is the result of the migration steps

easyCZ reviewed (commented):
LGTM % removal of team membership for admin user, which can break the dedicated sso setup flow

AlexTugarev commented on components/server/src/migration/user-to-team-migration-service.ts:
```diff
@@ -188,6 +188,14 @@ export class UserToTeamMigrationService {
     }
 
     needsMigration(user: User): boolean {
+        if (isBuiltinUser(user.id)) {
```
From PR description:
>  This path did not take into account the "single org" constraint, that's why login/setup flows started failing on Dedicated installations.

Can we _now_ make use of the `isSingleOrgMode` config here?

geropl replied:
> Can we now make use of the isSingleOrgMode config here?

We could, to guard createTeam, yes. :+1: But feels like outside of scope here.

AlexTugarev replied:
Was more pointing to disable `needsMigration` if `isSingleOrgMode` is true.

geropl replied:
I think tying to `!!user.organizationId` is more accurate, as that also applies for multi-tenant.

geropl commented on components/server/src/user/user-service.ts:
```diff
@@ -83,13 +83,7 @@ export class UserService {
         if (token) {
             await this.userDb.storeSingleToken(identity, token);
         }
-        // org-owned users have an org and the setup user should not get one automatically
```
Instead of doing some checks here and not there, these have now been moved into the migraterUser() method.

geropl replied:
Done.

geropl commented:
@easyCZ Re-requesting due to these changes:
 - now all users get `additionalData.isMigratedToTeamOnlyAttribution`, which seems to be more in line [with expectations](https://github.com/gitpod-io/gitpod/blob/35f3c98b980991d15cdfe30d5b82323efc0dd52d/components/server/src/user/user-service.ts#L88)
 - I adjusted, re-tested on staging (and fixed) the migrations

easyCZ reviewed (approved):
looks reasonable

geropl commented:
Will do one last test in a fresh preview, but then we can go with this.

geropl commented:
/unhold
